### ISSUE TITLE
`no-copy` flag unexpectedly removes `Clone` derive as well

### ISSUE THREAD

**Issue by divagant-martian** [rust-lang/rust-bindgen#2458]
### Input C/C++ Header

```C++
struct Bytes {
    char bytes[1024];
};
```

### Bindgen Invocation

```
$ bindgen input.h --no-copy Bytes
```

### Actual Results

```rust
/* automatically generated by rust-bindgen 0.64.0 */

#[repr(C)]
#[derive(Debug)]
pub struct Bytes {
    pub bytes: [::std::os::raw::c_char; 1024usize],
}
```

### Expected Results

By default bindgen adds `Copy` and `Clone` derives. Using `--no-copy` should only remove the `Copy` derivation. Types like these are too big to be `Copy` but are ok to be explicitly `Clone`d. TLDR:

```rust
/* automatically generated by rust-bindgen 0.64.0 */

#[repr(C)]
#[derive(Debug, Clone)]
pub struct Bytes {
    pub bytes: [::std::os::raw::c_char; 1024usize],
}
```

**Comment by pvdrz** [comment]
I agree this is a bit inflexible even though the behavior is documented. We need to:
- Update the docs for the `no_copy` method and the `--no-copy` flag explaining that both `Copy` and `Clone` are not derived. 
- Think how we can support removing `Copy` while keeping `Clone` in a backwards compatible manner.

I think finding a good name for this new option is going to be difficult as the clearer way would be a breaking change

## Current behavior
| Flags | Derives |
|-------|---------|
| None | `Copy`, `Clone` | 
| `--no-copy`||  

## Non-backwards compatible behavior
| Flags | Derives |
|-------|---------|
| None | `Copy`, `Clone` | 
| `--no-copy`|`Clone`| 
| `--no-clone`||  

## Backwards compatible behavior with not so good names
| Flags | Derives |
|-------|---------|
| None | `Copy`, `Clone` | 
| `--no-copy`|| 
| `--no-copy --keep-clone`|`Clone`| 

 Maybe we could introduce `--keep-clone` and add a warning when `--no-copy` is being used explaining that `--keep-clone` will be enabled by default in the future. Then we can introduce `--no-clone` and deprecate `--keep-clone` so everyone has enough time to fix their code.
 
 Pinging @emilio, @kulp and @amanjeev for feedback

**Comment by kulp** [comment]
> Maybe we could introduce `--keep-clone` and add a warning when `--no-copy` is being used explaining that `--keep-clone` will be enabled by default in the future. Then we can introduce `--no-clone` and deprecate `--keep-clone` so everyone has enough time to fix their code.

I think that solution would be reasonable, @pvdrz. I do not have a better idea.

<details>

This is just me working out what I think @pvdrz's last paragraph means.

## Phase 0 (current behavior)
| Flags | Derives |
|-------|---------|
| None | `Copy`, `Clone` | 
| `--no-copy`||  

## Phase 1
| Flags | Derives |
|-------|---------|
| None | `Copy`, `Clone` | 
| `--no-copy`| **yields deprecation warning about `--no-copy`** |
| `--no-copy --keep-clone`|`Clone`| 
| `--no-copy --no-clone`||

## Phase 2 (deferred until the next major release)
| Flags | Derives |
|-------|---------|
| None | `Copy`, `Clone` | 
| `--no-copy`| `Clone` |
| `--no-copy --keep-clone`|`Clone` + **yields deprecation warning about `--keep-clone`** | 
| `--no-copy --no-clone`||

## Phase 3 (steady state, possibly deferred until the subsequent major release)
| Flags | Derives |
|-------|---------|
| None | `Copy`, `Clone` | 
| `--no-copy`| `Clone` |
| `--no-copy --no-clone`||

</details>

**Comment by amanjeev** [comment]
thank you for insight!  I think introducing `--keep-clone` is a good resolution as long as 

- a warning/info is shown after someone uses `--no-copy` that "if they intended to keep clone they can use ..." (this is already suggested by @pvdrz )
- if possible, have `--keep-clone` be only used with `--no-copy` is used, otherwise has no effect.

**Comment by jschwe** [comment]
> I think finding a good name for this new option is going to be difficult as the clearer way would be a breaking change

I think it's important to differentiate between two different kinds of breaking changes
a) breaking change in behavior, but code keeps compiling (i.e. changeing the behavior of `--no-copy`). This is not good, since many users don't read changelogs (depandabot automerge).
b) breaking change, code stops compiling, with easy search and replace fix available: (e.g. rename `--no-copy` to `--no-clone` and and a new parameter `--no-copy2` for the behavior of deriving only `Clone` but not `Copy`. 

I think option b) is quite reasonable, since code will stop compiling, but users also have a very easy migration path. The name `--no-copy2` is of course horrible, but personally I'm also not a fan of making the API more complicated with `--keep-clone`.

**Comment by pvdrz** [comment]
@jschwe yeah I think option b is reasonable. Specially if we add an error when `--no-copy` is used explaining that you should use `--no-clone` or whatever. The only remaining issue is to find a good name for `--no-copy2`.
